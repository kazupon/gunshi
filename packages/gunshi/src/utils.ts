/**
 * @author kazuya kawaguchi (a.k.a. kazupon)
 * @license MIT
 */

import { ARG_PREFIX, BUILT_IN_KEY_SEPARATOR, BUILT_IN_PREFIX } from './constants.ts'

import type { Args } from 'args-tokens'
import type {
  Command,
  Commandable,
  CommandBuiltinArgsKeys,
  CommandBuiltinResourceKeys,
  CommandContext,
  CommandExamplesFetcher,
  DefaultGunshiParams,
  GenerateNamespacedKey,
  GunshiParams,
  KeyOfArgs,
  LazyCommand,
  RemovedIndex
} from './types.ts'

export { kebabnize } from 'args-tokens/utils'

export function isLazyCommand<G extends GunshiParams = DefaultGunshiParams>(
  cmd: unknown
): cmd is LazyCommand<G> {
  return typeof cmd === 'function' && 'commandName' in cmd && !!cmd.commandName
}

export async function resolveLazyCommand<G extends GunshiParams = DefaultGunshiParams>(
  cmd: Commandable<G>,
  name?: string | undefined,
  needRunResolving: boolean = false
): Promise<Command<G>> {
  let command: Command<G> | undefined
  if (isLazyCommand<G>(cmd)) {
    command = Object.assign(create<Command<G>>(), {
      name: cmd.commandName,
      description: cmd.description,
      args: cmd.args,
      examples: cmd.examples,
      resource: cmd.resource
    })

    if (needRunResolving) {
      const loaded = await cmd()
      if (typeof loaded === 'function') {
        command.run = loaded
      } else if (typeof loaded === 'object') {
        if (loaded.run == null) {
          throw new TypeError(`'run' is required in command: ${cmd.name || name}`)
        }
        command.run = loaded.run
        command.name = loaded.name
        command.description = loaded.description
        command.args = loaded.args
        command.examples = loaded.examples
        command.resource = loaded.resource
      } else {
        throw new TypeError(`Cannot resolve command: ${cmd.name || name}`)
      }
    }
  } else {
    command = Object.assign(create<Command<G>>(), cmd)
  }

  if (command.name == null && name) {
    command.name = name
  }

  return deepFreeze(command)
}

export function resolveBuiltInKey<
  K extends string = CommandBuiltinArgsKeys | CommandBuiltinResourceKeys
>(key: K): GenerateNamespacedKey<K> {
  return `${BUILT_IN_PREFIX}${BUILT_IN_KEY_SEPARATOR}${key}`
}

export function resolveArgKey<
  A extends Args = DefaultGunshiParams['args'],
  K extends string = KeyOfArgs<RemovedIndex<A>>
>(key: K): GenerateNamespacedKey<K, typeof ARG_PREFIX> {
  return `${ARG_PREFIX}${BUILT_IN_KEY_SEPARATOR}${key}`
}

export async function resolveExamples<G extends GunshiParams = DefaultGunshiParams>(
  ctx: Readonly<CommandContext<G>>,
  examples?: string | CommandExamplesFetcher<G>
): Promise<string> {
  return typeof examples === 'string'
    ? examples
    : typeof examples === 'function'
      ? await examples(ctx)
      : ''
}

export function mapResourceWithBuiltinKey(
  resource: Record<string, string>
): Record<string, string> {
  return Object.entries(resource).reduce((acc, [key, value]) => {
    acc[resolveBuiltInKey(key)] = value
    return acc
  }, create<Record<string, string>>())
}

export function create<T>(obj: object | null = null): T {
  return Object.create(obj) as T
}

export function log(...args: unknown[]): void {
  console.log(...args)
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function deepFreeze<T extends Record<string, any>>(obj: T): Readonly<T> {
  if (obj === null || typeof obj !== 'object') {
    return obj
  }

  for (const key of Object.keys(obj)) {
    const value = obj[key]
    if (typeof value === 'object' && value !== null) {
      deepFreeze(value)
    }
  }

  return Object.freeze(obj)
}
