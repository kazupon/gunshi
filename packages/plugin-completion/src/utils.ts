/**
 * @author kazuya kawaguchi (a.k.a. kazupon)
 * @license MIT
 */

function detectRuntime(): 'bun' | 'deno' | 'node' | 'unknown' {
  // @ts-ignore -- NOTE(kazupon): ignore, because `process` will detect ts compile error on `deno check`
  if (globalThis.process !== undefined && globalThis.process.release?.name === 'node') {
    return 'node'
  }
  // @ts-ignore -- NOTE(kazupon): ignore, because development env is node.js
  if (globalThis.Deno !== undefined) {
    return 'deno'
  }
  // @ts-ignore -- NOTE(kazupon): ignore, because development env is node.js
  if (globalThis.Bun !== undefined) {
    return 'bun'
  }
  return 'unknown'
}

function quoteIfNeeded(path: string): string {
  return path.includes(' ') ? `'${path}'` : path
}

/**
 * Quote the exec command for the current runtime.
 *
 * @returns The quoted exec command string
 */
export function quoteExec(): string {
  const runtime = detectRuntime()
  switch (runtime) {
    case 'node': {
      // @ts-ignore -- NOTE(kazupon): ignore, because `process` will detect ts compile error on `deno check`
      const execPath = globalThis.process.execPath
      // @ts-ignore -- NOTE(kazupon): ignore, because `process` will detect ts compile error on `deno check`
      const processArgs = globalThis.process.argv.slice(1)
      const quotedExecPath = quoteIfNeeded(execPath)
      // eslint-disable-next-line unicorn/no-array-callback-reference -- NOTE(kazupon): callback should be testable
      const quotedProcessArgs = processArgs.map(quoteIfNeeded)
      // @ts-ignore -- NOTE(kazupon): ignore, because `process` will detect ts compile error on `deno check`
      // eslint-disable-next-line unicorn/no-array-callback-reference -- NOTE(kazupon): callback should be testable
      const quotedProcessExecArgs = globalThis.process.execArgv.map(quoteIfNeeded)
      return `${quotedExecPath} ${quotedProcessExecArgs.join(' ')} ${quotedProcessArgs[0]}`
    }
    case 'deno': {
      throw new Error('deno not implemented yet, welcome contributions :)')
    }
    case 'bun': {
      throw new Error('bun not implemented yet, welcome contributions :)')
    }
    default: {
      throw new Error('Unsupported your javascript runtime for completion script generation.')
    }
  }
}
