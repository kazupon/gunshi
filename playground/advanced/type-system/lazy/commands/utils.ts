export function bundle({ target, minify }: { target: 'dev' | 'prod'; minify: boolean }) {
  console.log(`Building for ${target}...`)
  if (minify) {
    console.log('Minification enabled')
  }
}
