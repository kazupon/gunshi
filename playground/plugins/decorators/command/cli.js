import { cli, define } from 'gunshi'
import multi from './plugin.js'

const command = define({
  name: 'process',
  run: ctx => {
    console.log('>>> Executing actual command <<<')
    return 'Command result'
  }
})

await cli(process.argv.slice(2), command, {
  plugins: [multi]
})
