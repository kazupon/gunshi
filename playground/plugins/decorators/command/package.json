{
  "name": "command",
  "private": true,
  "packageManager": "pnpm@10.18.2",
  "type": "module",
  "main": "cli.js",
  "scripts": {
    "start": "node cli.js"
  },
  "dependencies": {
    "gunshi": "file:../../../../packages/gunshi"
  }
}
