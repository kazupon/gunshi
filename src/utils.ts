import { BUILT_IN_KEY_SEPARATOR, BUILT_IN_PREFIX } from './constants.ts'

import type { ArgOptions } from 'args-tokens'
import type {
  Command,
  Commandable,
  CommandBuiltinOptionsKeys,
  CommandBuiltinResourceKeys,
  GenerateNamespacedKey
} from './types.ts'

export async function resolveLazyCommand<Options extends ArgOptions = ArgOptions>(
  cmd: Commandable<Options>,
  name: string | undefined,
  entry: boolean = false
): Promise<Command<Options>> {
  const resolved = Object.assign(
    create<Command<Options>>(),
    typeof cmd == 'function' ? await cmd() : cmd,
    { default: entry }
  )

  if (resolved.name == null && name) {
    resolved.name = name
  }
  return deepFreeze(resolved)
}

export function resolveBuiltInKey<
  Key extends string = CommandBuiltinOptionsKeys | CommandBuiltinResourceKeys
>(key: Key): GenerateNamespacedKey<Key> {
  return `${BUILT_IN_PREFIX}${BUILT_IN_KEY_SEPARATOR}${key}`
}

export function mapResourceWithBuiltinKey(
  resource: Record<string, string>
): Record<string, string> {
  return Object.entries(resource).reduce((acc, [key, value]) => {
    acc[resolveBuiltInKey(key)] = value
    return acc
  }, create<Record<string, string>>())
}

export function create<T>(obj: object | null = null): T {
  return Object.create(obj) as T
}

export function log(...args: unknown[]): void {
  console.log(...args)
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function deepFreeze<T extends Record<string, any>>(obj: T): Readonly<T> {
  if (obj === null || typeof obj !== 'object') {
    return obj
  }

  for (const key of Object.keys(obj)) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const value = obj[key]
    if (typeof value === 'object' && value !== null) {
      deepFreeze(value)
    }
  }

  return Object.freeze(obj)
}
