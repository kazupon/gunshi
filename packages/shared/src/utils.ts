/**
 * @author kazuya kawaguchi (a.k.a. kazupon)
 * @license MIT
 */

import { kebabnize } from 'gunshi/utils'
import { ARG_PREFIX, BUILT_IN_KEY_SEPARATOR, BUILT_IN_PREFIX, PLUGIN_PREFIX } from './constants.ts'

import type {
  Args,
  ArgSchema,
  CommandContext,
  CommandExamplesFetcher,
  DefaultGunshiParams,
  GunshiParamsConstraint
} from 'gunshi'
import type {
  CommandBuiltinArgsKeys,
  CommandBuiltinResourceKeys,
  GenerateNamespacedKey,
  KeyOfArgs,
  RemovedIndex
} from './types.ts'

export function resolveBuiltInKey<
  K extends string = CommandBuiltinArgsKeys | CommandBuiltinResourceKeys
>(key: K): GenerateNamespacedKey<K> {
  return `${BUILT_IN_PREFIX}${BUILT_IN_KEY_SEPARATOR}${key}`
}

export function resolveArgKey<
  A extends Args = DefaultGunshiParams['args'],
  K extends string = KeyOfArgs<RemovedIndex<A>>
>(key: K, ctx?: Readonly<CommandContext>): string {
  return `${ctx?.name ? `${ctx.name}${BUILT_IN_KEY_SEPARATOR}` : ''}${ARG_PREFIX}${BUILT_IN_KEY_SEPARATOR}${key}`
}

export function resolveKey<
  T extends Record<string, string> = {},
  K = keyof T extends string ? keyof T : string
>(key: K, ctx?: Readonly<CommandContext>): string {
  return `${ctx?.name ? `${ctx.name}${BUILT_IN_KEY_SEPARATOR}` : ''}${key}`
}

export async function resolveExamples<G extends GunshiParamsConstraint = DefaultGunshiParams>(
  ctx: Readonly<CommandContext<G>>,
  examples?: string | CommandExamplesFetcher<G>
): Promise<string> {
  return typeof examples === 'string'
    ? examples
    : typeof examples === 'function'
      ? await examples(ctx)
      : ''
}

export function namespacedId<K extends string>(
  id: K
): GenerateNamespacedKey<K, typeof PLUGIN_PREFIX> {
  return `${PLUGIN_PREFIX}${BUILT_IN_KEY_SEPARATOR}${id}`
}

export function makeShortLongOptionPair(
  schema: ArgSchema,
  name: string,
  toKebab?: boolean
): string {
  // Convert camelCase to kebab-case for display in help text if toKebab is true
  const displayName = toKebab || schema.toKebab ? kebabnize(name) : name
  let key = `--${displayName}`
  if (schema.short) {
    key = `-${schema.short}, ${key}`
  }
  return key
}
