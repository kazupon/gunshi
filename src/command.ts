import { parseArgs, resolveArgs } from 'args-tokens'
import { COMMAND_OPTIONS_DEFAULT, COMMON_OPTIONS } from './constants.js'
import { createCommandContext } from './context.js'
import {
  renderHeader,
  renderUsage,
  renderUsageDefault,
  renderValidationErrors
} from './renderer.js'
import { log, nullObject } from './utils.js'

import type { ArgOptions, ArgToken } from 'args-tokens'
import type {
  Command,
  CommandContext,
  CommandEnvironment,
  CommandOptions,
  CommandRunner,
  LazyCommand
} from './types'

/**
 * Run a command with given arguments and environment
 * @param args - command line arguments
 * @param env - a {@link CommandEnvironment | command environment}
 * @param opts - a {@link CommandOptions | command options}
 */
export async function gunshi<Options extends ArgOptions>(
  args: string[],
  envOrEntry: CommandEnvironment<Options> | CommandRunner<Options>,
  opts: CommandOptions<Options> = COMMAND_OPTIONS_DEFAULT
): Promise<void> {
  const tokens = parseArgs(args)

  const raw = getCommandRaw(tokens)
  const [name, command, env] = await resolveCommand(raw, envOrEntry)
  if (!command) {
    throw new Error(`Command not found: ${name || ''}`)
  }

  const options = resolveOptions(command.options)

  const { values, positionals, error } = resolveArgs(options, tokens)
  const ctx = createCommandContext(
    options,
    values,
    positionals,
    env,
    command,
    opts as Required<CommandOptions<Options>>
  )
  if (values.version) {
    showVersion(ctx)
    return
  }

  await showHeader(ctx)

  if (values.help) {
    if (raw) {
      await showUsage(ctx)
      return
    } else {
      // omitted command
      await showUsageDefault(ctx)
      return
    }
  }

  if (error) {
    await showValidationErrors(ctx, error)
    throw error
  }

  await command.run(ctx)
}

function resolveOptions<Options extends ArgOptions>(options?: Options): Options {
  return Object.assign(Object.create(null) as Options, options, COMMON_OPTIONS)
}

function getCommandRaw(tokens: ArgToken[]): string {
  const firstToken = tokens[0]
  return firstToken &&
    firstToken.kind === 'positional' &&
    firstToken.index === 0 &&
    firstToken.value
    ? firstToken.value
    : ''
}

async function showUsage<Options extends ArgOptions>(ctx: CommandContext<Options>): Promise<void> {
  if (ctx.commandOptions.renderUsage === null) {
    return
  }
  const render = ctx.commandOptions.renderUsage || renderUsage
  log(await render(ctx))
}

async function showUsageDefault<Options extends ArgOptions>(
  ctx: CommandContext<Options>
): Promise<void> {
  if (ctx.commandOptions.renderUsageDefault === null) {
    return
  }
  const render = ctx.commandOptions.renderUsageDefault || renderUsageDefault
  log(await render(ctx))
}

function showVersion<Options extends ArgOptions>(ctx: CommandContext<Options>): void {
  log(ctx.env.version)
}

async function showHeader<Options extends ArgOptions>(ctx: CommandContext<Options>): Promise<void> {
  if (ctx.commandOptions.renderHeader === null) {
    return
  }
  const header = await (ctx.commandOptions.renderHeader || renderHeader)(ctx)
  if (header) {
    log(header)
    log()
  }
}

async function showValidationErrors<Options extends ArgOptions>(
  ctx: CommandContext<Options>,
  error: AggregateError
): Promise<void> {
  if (ctx.commandOptions.renderValidationErrors === null) {
    return
  }
  const render = ctx.commandOptions.renderValidationErrors || renderValidationErrors
  log(await render(ctx, error))
}

async function resolveCommand<Options extends ArgOptions>(
  raw: string,
  envOrEntry: CommandEnvironment<Options> | CommandRunner<Options>
): Promise<[string | undefined, Command<Options> | undefined, CommandEnvironment<Options>]> {
  const omitted = !raw
  if (typeof envOrEntry === 'function') {
    const cmd = { run: envOrEntry } satisfies Command<Options>
    return [undefined, cmd, { entry: cmd }]
  } else {
    if (omitted) {
      let name: string | undefined
      if (envOrEntry.entry) {
        if (typeof envOrEntry.entry === 'string') {
          name = envOrEntry.entry
        } else if (typeof envOrEntry.entry === 'object') {
          return [envOrEntry.entry.name, envOrEntry.entry, envOrEntry]
        }
      }

      // eslint-disable-next-line unicorn/no-null
      if (envOrEntry.subCommands == null) {
        return [undefined, undefined, envOrEntry]
      }

      if (name) {
        // find sub command with entry command name
        return [raw, await loadCommand(raw, envOrEntry), envOrEntry]
      } else {
        // find command from such commands that has default flag
        const loaded = await Promise.all(
          Object.entries(envOrEntry.subCommands || nullObject()).map(
            async ([_, cmd]) => await resolveLazyCommand(cmd)
          )
        )
        const found = loaded.find(cmd => cmd.default)
        return found ? [found.name, found, envOrEntry] : [undefined, undefined, envOrEntry]
      }
    } else {
      // eslint-disable-next-line unicorn/no-null
      if (envOrEntry.subCommands == null) {
        return [raw, undefined, envOrEntry]
      }
      return [raw, await loadCommand(raw, envOrEntry), envOrEntry]
    }
  }
}

async function resolveLazyCommand<Options extends ArgOptions>(
  cmd: Command<Options> | LazyCommand<Options>
): Promise<Command<Options>> {
  return typeof cmd == 'function' ? await cmd() : cmd
}

async function loadCommand<Options extends ArgOptions>(
  name: string,
  env: CommandEnvironment<Options>
): Promise<Command<Options>> {
  const cmd = env.subCommands![name]
  return await resolveLazyCommand(cmd)
}
