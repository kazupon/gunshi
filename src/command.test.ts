import { afterEach, describe, expect, test, vi } from 'vitest'
import { defineMockLog } from '../test/utils'
import { gunshi } from './command'

import type { ArgOptions } from 'args-tokens'
import type { Command, LazyCommand } from './types'

afterEach(() => {
  vi.resetAllMocks()
})

describe('execute command', () => {
  test('entry iniline function', async () => {
    const mockFn = vi.fn()
    await gunshi([], mockFn)
    expect(mockFn).toBeCalled()
  })

  test('entry command', async () => {
    const mockFn = vi.fn()
    await gunshi([], {
      run: mockFn
    })
    expect(mockFn).toBeCalled()
  })

  test('entry strictly command + sub commands', async () => {
    const mockShow = vi.fn()
    const mockCommand1 = vi.fn()
    const mockCommand2 = vi.fn()
    const show = {
      name: 'show',
      run: mockShow
    }
    const subCommands = new Map<string, Command<ArgOptions> | LazyCommand<ArgOptions>>()
    subCommands.set('command1', {
      name: 'command1',
      run: mockCommand1
    })
    subCommands.set('command2', {
      name: 'command2',
      run: mockCommand2
    })
    const options = {
      subCommands
    }

    await gunshi([''], show, options) // omit
    await gunshi(['show'], show, options)
    await gunshi(['command1'], show, options)
    await gunshi(['command2'], show, options)

    expect(mockShow).toBeCalledTimes(2)
    expect(mockCommand1).toBeCalledTimes(1)
    expect(mockCommand2).toBeCalledTimes(1)
  })

  test('entry loose command + sub commands', async () => {
    const mockAnonymous = vi.fn()
    const mockShow = vi.fn()
    const mockCommand1 = vi.fn()
    const mockCommand2 = vi.fn()
    // no name command
    const anonymous = {
      run: mockAnonymous
    }
    const subCommands = new Map<string, Command<ArgOptions> | LazyCommand<ArgOptions>>()
    subCommands.set('show', {
      run: mockShow
    })
    subCommands.set('command1', {
      name: 'command1',
      run: mockCommand1
    })
    subCommands.set('command2', {
      name: 'command2',
      run: mockCommand2
    })
    const options = {
      subCommands
    }

    await gunshi([''], anonymous, options) // omit
    await gunshi(['show'], anonymous, options)
    await gunshi(['command1'], anonymous, options)
    await gunshi(['command2'], anonymous, options)

    expect(mockAnonymous).toBeCalledTimes(1)
    expect(mockShow).toBeCalledTimes(1)
    expect(mockCommand1).toBeCalledTimes(1)
    expect(mockCommand2).toBeCalledTimes(1)
  })

  test('command not found', async () => {
    await expect(async () => {
      await gunshi(['show'], { run: vi.fn() }, {})
    }).rejects.toThrowError('Command not found: show')
  })
})

describe('aute generate usage', () => {
  test('inline function', async () => {
    const utils = await import('../src/utils')
    const log = defineMockLog(utils)
    await gunshi(['-h'], vi.fn())

    const message = log()
    expect(message).toMatchSnapshot()
  })

  test('loosely entry command', async () => {
    const utils = await import('../src/utils')
    const log = defineMockLog(utils)
    await gunshi(['-h'], {
      options: {
        foo: {
          type: 'string',
          short: 'f'
        }
      },
      run: vi.fn()
    })

    const message = log()
    expect(message).toMatchSnapshot()
  })

  test('strictly entry command', async () => {
    const utils = await import('../src/utils')
    const log = defineMockLog(utils)
    await gunshi(
      ['-h'],
      {
        options: {
          foo: {
            type: 'string',
            short: 'f'
          }
        },
        name: 'command1',
        usage: {
          options: {
            foo: 'The foo option'
          },
          examples: '# Example 1\n$ gunshi --foo bar\n# Example 2\n$ gunshi -f bar'
        },
        run: vi.fn()
      },
      {
        name: 'gunshi',
        description: 'Modern CLI tool',
        version: '0.0.0',
        usageOptionType: true
      }
    )

    const message = log()
    expect(message).toMatchSnapshot()
  })

  test('loosely sub commands', async () => {
    const utils = await import('../src/utils')
    const log = defineMockLog(utils)

    const entryOptions = {
      foo: {
        type: 'string',
        short: 'f'
      }
    } satisfies ArgOptions
    const entry = {
      options: entryOptions,
      run: vi.fn()
    } satisfies Command<typeof entryOptions>

    const command2Options = {
      bar: {
        type: 'number',
        short: 'b',
        default: 42
      }
    } satisfies ArgOptions
    const command2 = {
      options: command2Options,
      run: vi.fn()
    } satisfies Command<typeof command2Options>

    type CommandArgs = typeof entryOptions | typeof command2Options
    const subCommands = new Map<string, Command<CommandArgs> | LazyCommand<CommandArgs>>()
    subCommands.set('command2', command2)

    await gunshi(['-h'], entry, { subCommands })
    await gunshi(['command2', '-h'], entry, { subCommands })

    const message = log()
    expect(message).toMatchSnapshot()
  })

  test('strictly sub commands', async () => {
    const utils = await import('../src/utils')
    const log = defineMockLog(utils)

    const entryOptions = {
      foo: {
        type: 'string',
        short: 'f'
      }
    } satisfies ArgOptions
    const entry = {
      options: entryOptions,
      name: 'command1',
      usage: {
        options: {
          foo: 'The foo option'
        },
        examples: '# Example 1\n$ gunshi --foo bar\n# Example 2\n$ gunshi -f bar'
      },
      run: vi.fn()
    } satisfies Command<typeof entryOptions>

    const command2Options = {
      bar: {
        type: 'number',
        short: 'b',
        default: 42
      }
    } satisfies ArgOptions
    const command2 = {
      options: command2Options,
      name: 'command2',
      usage: {
        options: {
          bar: 'The bar option'
        },
        examples: '# Example 1\n$ gunshi command2 --bar 42\n# Example 2\n$ gunshi command2 -b 42'
      },
      run: vi.fn()
    } satisfies Command<typeof command2Options>

    type CommandArgs = typeof entryOptions | typeof command2Options
    const subCommands = new Map<string, Command<CommandArgs> | LazyCommand<CommandArgs>>()
    subCommands.set('command2', command2)

    await gunshi(['-h'], entry, {
      subCommands,
      name: 'gunshi',
      description: 'Modern CLI tool',
      version: '0.0.0',
      leftMargin: 4,
      middleMargin: 15
    })

    await gunshi(['command2', '-h'], entry, {
      subCommands,
      name: 'gunshi',
      description: 'Modern CLI tool',
      version: '0.0.0'
    })

    const message = log()
    expect(message).toMatchSnapshot()
  })
})

describe.todo('custom generate usage', () => {})
