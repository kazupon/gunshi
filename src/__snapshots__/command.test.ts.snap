// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`aute generate usage > inline function 1`] = `
"USAGE:
  COMMAND <OPTIONS> 

OPTIONS:
  -h, --help             Display this help message
  -v, --version          Display this version
"
`;

exports[`aute generate usage > loosely entry command 1`] = `
"USAGE:
  COMMAND <OPTIONS> 

OPTIONS:
  -f, --foo <foo>          
  -h, --help               Display this help message
  -v, --version            Display this version
"
`;

exports[`aute generate usage > loosely sub commands 1`] = `
"USAGE:
  COMMAND <OPTIONS> 

OPTIONS:
  -f, --foo <foo>          
  -h, --help               Display this help message
  -v, --version            Display this version

USAGE:
  COMMAND command2 <OPTIONS>

OPTIONS:
  -b, --bar [bar]          
  -h, --help               Display this help message
  -v, --version            Display this version
"
`;

exports[`aute generate usage > strictly entry command 1`] = `
"Modern CLI tool (gunshi v0.0.0)

USAGE:
  gunshi <OPTIONS> 

OPTIONS:
  -f, --foo <foo>          [string]  The foo option
  -h, --help               [boolean] Display this help message
  -v, --version            [boolean] Display this version

EXAMPLES:
  # Example 1
  $ gunshi --foo bar
  # Example 2
  $ gunshi -f bar
"
`;

exports[`aute generate usage > strictly sub commands 1`] = `
"Modern CLI tool (gunshi v0.0.0)

USAGE:
    gunshi [command1] <OPTIONS> 
    gunshi <COMMANDS>

COMMANDS:
    command2                 
    command1                 

For more info, run any command with the \`--help\` flag:
    gunshi command2 --help
    gunshi command1 --help

OPTIONS:
    -f, --foo <foo>               The foo option
    -h, --help                    Display this help message
    -v, --version                 Display this version

EXAMPLES:
    # Example 1
    $ gunshi --foo bar
    # Example 2
    $ gunshi -f bar

Modern CLI tool (gunshi v0.0.0)

USAGE:
  gunshi command2 <OPTIONS>

COMMANDS:
  command2            
  command1            

For more info, run any command with the \`--help\` flag:
  gunshi command2 --help
  gunshi command1 --help

OPTIONS:
  -b, --bar [bar]          The bar option
  -h, --help               Display this help message
  -v, --version            Display this version

EXAMPLES:
  # Example 1
  $ gunshi command2 --bar 42
  # Example 2
  $ gunshi command2 -b 42
"
`;
